In [1]:
import os
import getpass
import requests
API_KEY = getpass.getpass()


··········


In [1]:
# import os
# os.getcwd()

'/content'

In [2]:
import google.generativeai as genai

genai.configure(api_key=API_KEY)

In [3]:
for m in genai.list_models():
  if 'generateContent' in m.supported_generation_methods:
    print(m.name)

models/gemini-1.0-pro
models/gemini-1.0-pro-001
models/gemini-1.0-pro-latest
models/gemini-1.0-pro-vision-latest
models/gemini-1.5-pro-latest
models/gemini-pro
models/gemini-pro-vision


In [4]:
model = genai.GenerativeModel('gemini-pro')


In [9]:
from IPython.display import display
from IPython.display import Markdown
import pathlib
import textwrap
def to_markdown(text):
  text = text.replace('•', '  *')
  return Markdown(textwrap.indent(text, '> ', predicate=lambda _: True))

In [10]:

prompt = f"Generate a travel itinerary for a city destination with a budget of $1000, " \
             f"from Apr 1st to Apr 8th, including activities like city walking, visiting museums. ."
response = model.generate_content(prompt)
to_markdown(response.text)


> **Day 1 (April 1st)**
> 
> * Arrive and check into a budget-friendly hostel ($25/night for 7 nights: $175)
> * Stroll through the historic district, admiring the architecture and browsing souvenir shops.
> * Visit the City Museum (Free admission)
> 
> **Day 2 (April 2nd)**
> 
> * Explore the vibrant food scene with a guided street food tour ($50)
> * Visit the National Art Museum (Free admission)
> * Take a scenic boat cruise along the river ($25)
> 
> **Day 3 (April 3rd)**
> 
> * Wander through the botanical gardens, exploring exotic plant life. (Free admission)
> * Visit the Science and Technology Museum ($20)
> * Enjoy a dinner at a local restaurant ($20)
> 
> **Day 4 (April 4th)**
> 
> * Climb to the top of the observation tower for panoramic city views ($15)
> * Walk through the quaint streets, discovering hidden gems.
> * Visit the History Museum ($10)
> 
> **Day 5 (April 5th)**
> 
> * Explore the colorful markets, bargaining for local crafts and souvenirs.
> * Visit the Natural History Museum ($15)
> * Grab dinner at a street vendor ($10)
> 
> **Day 6 (April 6th)**
> 
> * Visit the Zoo ($25)
> * Attend a cultural performance at a local theater ($30)
> * Enjoy a farewell dinner at a traditional restaurant ($30)
> 
> **Day 7 (April 7th)**
> 
> * Do some last-minute shopping for souvenirs and gifts.
> * Visit the planetarium for an immersive celestial experience ($20)
> 
> **Day 8 (April 8th)**
> 
> * Depart from the city
> 
> **Estimated Expenses:**
> 
> * Accommodation: $175
> * Activities: $190
> * Food and drinks: $150
> * Transportation: $25
> * Shopping: $50
> * Souvenirs: $50
> * Contingency fund: $60
> 
> **Total:** $950
> 
> This itinerary provides a balance of cultural, historical, and entertainment activities while staying within your budget of $1,000.

## Write Clear Instructions


In [11]:
prompt = "Create a detailed travel itinerary for a city destination within a budget of $1000, for the period from April 1st to April 8th. The itinerary should include activities such as city walking and visiting museums. Please ensure that the schedule is realistic and includes costs for each activity."

In [12]:
response = model.generate_content(prompt)
to_markdown(response.text)

> **Day 1 (April 1st)**
> 
> * Arrive in the city and check into a budget-friendly hotel ($50/night)
> * Stroll through the city's downtown area (Free)
> * Visit the city's main market for souvenirs and local goods ($20)
> * Enjoy dinner at a local restaurant ($15)
> 
> **Day 2 (April 2nd)**
> 
> * Embark on a guided walking tour of the city's historical landmarks ($25)
> * Explore the city's art district and visit a free art gallery (Free)
> * Have lunch at a street food stall ($10)
> * Visit a local park for a relaxing afternoon ($5 for entry)
> 
> **Day 3 (April 3rd)**
> 
> * Dedicate the day to museum exploration:
>     * Visit the city's natural history museum ($15)
>     * Explore the local science museum ($12)
> 
> **Day 4 (April 4th)**
> 
> * Take a day trip to a nearby attraction:
>     * Visit a historical village or castle ($20 for transportation and entry)
>     * Go for a hike in a scenic park (Free)
> 
> **Day 5 (April 5th)**
> 
> * Learn about the city's history and culture at its local history museum ($10)
> * Visit a local university and explore its campus (Free)
> * Have dinner at a fine dining restaurant for a special night out ($30)
> 
> **Day 6 (April 6th)**
> 
> * Go shopping in the city's main shopping district ($50 for souvenirs)
> * Enjoy a relaxing spa treatment ($60)
> * Have dinner at a restaurant with live music ($20)
> 
> **Day 7 (April 7th)**
> 
> * Visit a zoo or aquarium ($25)
> * Take a boat tour of the city's waterways ($20)
> * Have a farewell dinner at a local restaurant ($15)
> 
> **Day 8 (April 8th)**
> 
> * Depart the city
> 
> **Budget Summary:**
> 
> * Hotel: $200
> * Activities: $160
> * Food and Drinks: $125
> * Transportation: $40
> * Souvenirs: $70
> * Spa Treatment: $60
> 
> **Total Estimated Cost: $955**

## Provide Reference Text


In [13]:
prompt = 'Using the example of a previous successful itinerary: "Day 1: Morning - city walk ($0), Afternoon - visit the National Museum ($15), Evening - local cuisine dinner ($30)", generate a full travel itinerary for a city destination from April 1st to April 8th with a budget of $1000. Include similar activities like city walking and visiting museums, and provide costs for each activity.'

In [14]:
response = model.generate_content(prompt)
to_markdown(response.text)

> **Day 1**
> 
> * Morning: City walking tour (free)
> * Afternoon: Visit the National Museum ($15)
> * Evening: Local cuisine dinner ($30)
> 
> **Day 2**
> 
> * Morning: Explore the city's historic district (free)
> * Afternoon: Visit the Art Museum ($20)
> * Evening: Dinner at a mid-range restaurant ($40)
> 
> **Day 3**
> 
> * Morning: Visit a local park or botanical garden (free)
> * Afternoon: Shopping for souvenirs ($50)
> * Evening: Street food crawl ($25)
> 
> **Day 4**
> 
> * Morning: Visit the city zoo ($25)
> * Afternoon: Take a guided bus tour ($30)
> * Evening: Dinner at a rooftop restaurant ($50)
> 
> **Day 5**
> 
> * Morning: Attend a cooking class ($60)
> * Afternoon: Free time for relaxation
> * Evening: Farewell dinner at a local restaurant ($40)
> 
> **Day 6**
> 
> * Morning: Depart for home
> 
> **Estimated Cost Breakdown:**
> 
> * Activities: $185
> * Food and drinks: $245
> * Accommodation (6 nights at $100/night): $600
> * Transportation (airport transfers/taxis): $70
> 
> **Total:** $1000

## Split Complex Tasks into Simpler Subtasks

In [15]:
prompt = """
Step 1: List potential cities that can be visited on a $1000 budget.
Step 2: For the selected city, propose a daily schedule from April 1st to April 8th.
Step 3: Include activities such as city walking and museum visits. Provide entry costs for each museum.
Step 4: Sum up the costs for each day to ensure the total does not exceed $1000.

"""

In [16]:
response = model.generate_content(prompt)
to_markdown(response.text)

> **Step 1: Potential cities for a $1000 budget**
> 
> * Bangkok, Thailand
> * Istanbul, Turkey
> * Budapest, Hungary
> * Prague, Czech Republic
> * Krakow, Poland
> 
> **Step 2: Selected city - Bangkok, Thailand**
> 
> **Step 3: Daily schedule**
> 
> **April 1st**
> 
> * Arrive in Bangkok and check into a budget-friendly hotel near the city center. (Cost: $20/night)
> * Visit the Grand Palace and Wat Phra Kaew (Temple of the Emerald Buddha). (Cost: 500 baht / $15)
> * Take a walk along Khaosan Road, a famous backpacker area. (Cost: Free)
> 
> **Total cost: $35**
> 
> **April 2nd**
> 
> * Visit the Jim Thompson House, a beautiful silk museum. (Cost: 200 baht / $6)
> * Explore the Wat Pho Temple and receive a traditional Thai massage. (Cost: 300 baht / $9)
> * Take a boat trip along the Chao Phraya River. (Cost: 40 baht / $1.20)
> 
> **Total cost: $16.20**
> 
> **April 3rd**
> 
> * Visit the Museum of Siam, which tells the story of Thailand's history and culture. (Cost: 300 baht / $9)
> * Explore the Chatuchak Weekend Market, one of the largest markets in the world. (Cost: Free)
> 
> **Total cost: $9**
> 
> **April 4th**
> 
> * Take a day trip to Ayutthaya, an ancient city with many historical ruins. (Cost: 800 baht / $24, including transportation and entrance fees)
> 
> **Total cost: $24**
> 
> **April 5th**
> 
> * Visit the Bangkok Art and Culture Centre, which features contemporary art exhibitions. (Cost: Free)
> * Walk around Lumpini Park, a large green space in the city center. (Cost: Free)
> 
> **Total cost: $0**
> 
> **April 6th**
> 
> * Visit the National Museum of Bangkok, which houses a collection of Thai art and artifacts. (Cost: 200 baht / $6)
> * Explore the Banglamphu neighborhood, known for its street food and night market. (Cost: Varies depending on purchases)
> 
> **Total cost: $6**
> 
> **April 7th**
> 
> * Visit the Wat Arun (Temple of Dawn) and enjoy the views from the temple's terrace. (Cost: 50 baht / $1.50)
> * Take a cooking class and learn how to prepare traditional Thai dishes. (Cost: 1,000 baht / $30)
> 
> **Total cost: $31.50**
> 
> **April 8th**
> 
> * Depart from Bangkok.
> 
> **Step 4: Costs summary**
> 
> * Hotel: $20 x 7 nights = $140
> * Museums/Temples: $15 + $6 + $9 + $6 + $9 + $6 + $1.50 + $30 = $82.50
> * Activities: $16.20 + $24 + $0 + $31.50 = $71.70
> * Transportation: $800 (day trip to Ayutthaya)
> * Food and other expenses: $206.80
> * 1000 - 140 - 82.50 - 71.70 - 800 - 206.80 = $0.00
> 
> **Total cost: $1000**

## Give the Model Time to "Think"


In [17]:
prompt = 'Please take a moment to consider the best options. Then, generate a travel itinerary for a city destination within a $1000 budget from April 1st to April 8th. The itinerary should focus on low-cost activities like city walking and visiting museums, including the cost for each activity.'

In [18]:
response = model.generate_content(prompt)
to_markdown(response.text)

> **Destination:** Berlin, Germany
> 
> **Budget:** $1000
> 
> **Itinerary:**
> 
> **Day 1 (April 1st)**
> 
> * Arrive in Berlin and check into a budget-friendly hostel (approx. $25/night)
> * Stroll through the Brandenburg Gate (free)
> * Explore the Reichstag Building (free guided tour)
> * Enjoy a traditional German dinner at a local restaurant (approx. $20)
> 
> **Day 2 (April 2nd)**
> 
> * Visit the Pergamon Museum (entrance fee: $12)
> * Walk along the Berlin Wall Memorial (free)
> * Have lunch at a street food stand (approx. $10)
> * Take a guided tour of the Topography of Terror (entrance fee: $8)
> 
> **Day 3 (April 3rd)**
> 
> * Discover the Gemäldegalerie (entrance fee: $10)
> * Explore the Jewish Museum (entrance fee: $12)
> * Have dinner at a budget-friendly restaurant (approx. $15)
> 
> **Day 4 (April 4th)**
> 
> * Visit the Berlin Cathedral (free admission)
> * Walk through the Tiergarten park (free)
> * Have lunch at a picnic spot (approx. $5)
> * Take a boat tour along the Spree River (approx. $15)
> 
> **Day 5 (April 5th)**
> 
> * Explore the East Side Gallery (free)
> * Visit the German Historical Museum (entrance fee: $8)
> * Have dinner at a restaurant with live music (approx. $20)
> 
> **Day 6 (April 6th)**
> 
> * Visit the Technical Museum (entrance fee: $12)
> * Take a walk through the Mauerpark flea market (free)
> * Have lunch at a market stall (approx. $5)
> * Enjoy an evening concert at a local bar (approx. $10)
> 
> **Day 7 (April 7th)**
> 
> * Visit the Charlottenburg Palace (entrance fee: $10)
> * Take a guided tour of the Berliner Unterwelten (underground tunnels) (approx. $15)
> * Have dinner at a traditional biergarten (approx. $15)
> 
> **Day 8 (April 8th)**
> 
> * Depart from Berlin
> 
> **Estimated Costs:**
> 
> * Accommodation: $175
> * Entrance fees: $75
> * Food and drinks: $150
> * Transportation and activities: $50
> * Contingency fund: $50
> 
> **Total:** $400
> 
> **Note:** This itinerary allows for flexibility in budgeting for food and activities, as well as unforeseen expenses. Costs may vary based on individual preferences and the availability of discounts.

## Use External Tools


In [21]:
prompt = 'Use a budget planning tool to calculate daily expenses. Create a travel itinerary for a city destination from April 1st to April 8th, staying under a $1000 budget. Include activities such as city walking and visiting museums, ensuring to list and sum the costs associated with each activity.'
response = model.generate_content(prompt)
response.text


"**Budget Planning Tool**\n\n**Expense Category | Daily Amount | Daily Total | Weekly Total**\n---|---|---|---|\nAccommodation | $50 | $350 | $350\nFood & Drinks | $25 | $175 | $175\nTransportation | $10 | $70 | $70\nActivities | $15 | $105 | $105\nContingency | $10 | $70 | $70\n**Total** | | | **$770**\n\n**Travel Itinerary**\n\n**Day 1 (April 1st)**\n\n* Arrive at the city and check into a budget-friendly hostel: $50\n* City walking tour: Free\n* Visit the National Museum of Art: $15\n\n**Day 2 (April 2nd)**\n\n* Visit the City History Museum: $10\n* Explore the local market and have lunch: $20\n* Take a boat tour along the river: $20\n\n**Day 3 (April 3rd)**\n\n* Visit the Science and Technology Museum: $15\n* Have a dinner at a local restaurant: $30\n\n**Day 4 (April 4th)**\n\n* Attend a free city concert in the park\n* Visit the Museum of Contemporary Art: $10\n* Have dinner at a street food market: $15\n\n**Day 5 (April 5th)**\n\n* Visit the Botanical Gardens: Free\n* Have a picn

### Result:

**Budget Planning Tool**

Expense Category | Daily Amount | Daily Total | Weekly Total
---|---|---|---|
Accommodation | \$50 | \$350 | \$350
Food & Drinks | \$25 | \$175 | \$175
Transportation | \$10 | \$70 | \$70
Activities | \$15 | \$105 | \$105
Contingency | \$10 | \$70 | \$70
Total | | | **\$770**

**Travel Itinerary**

**Day 1 (April 1st)**

* Arrive at the city and check into a budget-friendly hostel: $50
* City walking tour: Free
* Visit the National Museum of Art: $15

**Day 2 (April 2nd)**

* Visit the City History Museum: $10
* Explore the local market and have lunch: $20
* Take a boat tour along the river: $20

**Day 3 (April 3rd)**

* Visit the Science and Technology Museum: $15
* Have a dinner at a local restaurant: $30

**Day 4 (April 4th)**

* Attend a free city concert in the park
* Visit the Museum of Contemporary Art: $10
* Have dinner at a street food market: $15

**Day 5 (April 5th)**

* Visit the Botanical Gardens: Free
* Have a picnic lunch: $10
* Take a guided tour of the city's historical district: $15

**Day 6 (April 6th)**

* Visit the Zoological Park: $15
* Explore the city's nightlife scene: $30

**Day 7 (April 7th)**

* Visit the Aquarium: $20
* Have a farewell dinner at a rooftop restaurant: $40

**Day 8 (April 8th)**

* Depart from the city

**Cost Summary**

* Accommodation: \$350
* Food & Drinks: \$175
* Transportation: \$70
* Activities: \$105
* Contingency: \$70

Total: \$770

This itinerary provides a comprehensive travel experience while staying within the budget of $1000. By choosing budget-friendly accommodation, exploring free attractions, and dining at local markets, you can fully immerse yourself in the city's culture without breaking the bank.
